### DISCUSSION TITLE
tutorials/3d/using_transforms

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#243] [User-contributed notes]
# tutorials/3d/using_transforms

Introduction: If you have never made 3D games before, working with rotations in three dimensions can be confusing at first. Coming from 2D, the natural way of thinking is along the lines of"Oh, it'...

https://docs.godotengine.org/en/stable/tutorials/3d/using_transforms.html

**Comment by StealthFanAttic** [comment]
Great article and apart from the camera shake code which sets the rotation property rather than using a Transform, the Third Person Shooter demo code is a nice supplement to this.

Just clarifying, the FPS camera code under the Setting information section rotates in Y then in X, is this because the default Euler Rotation order in a Node3D is YXZ?

**Comment by TrevorBlythe** [comment]
"Interpolating between two transforms can efficiently be done with quaternions."
Is it really more efficient if you have to go through all that overhead? If transforms are how things are stored natively in the engine, I see no reason to convert it to something else just to have to convert it back.

**Reply by pennyloafers** [reply]
Yes, because of 3x3 matrix multiplication is a lot more intensive then quaternion math.

**Reply by pennyloafers** [reply]
It also allows you to interpolate with slerp which a basis alone cannot achieve.

**Reply by rozab** [reply]
Basis does have a slerp method, it just does the Quaternion slerp and then a lerp to account for scale. Probably a smaller performance hit than doing this in gdscript (don't quote me). Teaching it this way probably makes sense though.

https://github.com/godotengine/godot/blob/c81fd6c51233a727da528cf7f74137d56b5d6efe/core/math/basis.cpp#L910

**Comment by Joy-less** [comment]
> ```cs
> // Convert basis to quaternion, keep in mind scale is lost
> var a = transform.Basis.GetQuaternion();
> var b = transform2.Basis.GetQuaternion();
> // Interpolate using spherical-linear interpolation (SLERP).
> var c = a.Slerp(b, 0.5f); // find halfway point between a and b
> // Apply back
> transform.Basis = new Basis(c);
> ```

`GetQuaternion` should be `GetRotationQuaternion`.
See https://github.com/godotengine/godot/pull/55675

**Reply by raulsntos** [reply]
Thanks. Opened a PR to sync the C# example with GDScript:

- https://github.com/godotengine/godot-docs/pull/10808

**Comment by Kuvaitt** [comment]
I've been trying to make a security camera look towards a target and clamp its orientation. Trying to figure out whether I should use rotations or transforms (and whether they need to be in local and global space) is driving me insane!

If I'm supposed to only use transforms, without getting/setting the rotation of an object, how the hell am I supposed to clamp an object's rotation (with the clamp values being defined in degrees)? How do you clamp a transform?

**Reply by rozab** [reply]
I prefer to work in degrees. I have these helper functions I use to calculate the rotation, then I can clamp and apply it:

```
static func pitch_angle_to(to: Vector3) -> float:
	return atan2(to.y, -to.z)

static func yaw_angle_to(to: Vector3) -> float:
	return atan2(-to.x, -to.z)
```

If you're doing this, have the camera be the child of another object which you don't rotate. This 'remembers' the starting point, then you can set the local transform of the child each time.

**Reply by Kuvaitt** [reply]
@rozab Yeah, I also had something like this:
```
public static Vector3 GetRotationTowardTarget(Transform3D transform, Vector3 targetPosition) {
	return transform.LookingAt(targetPosition, Vector3.Up).Basis.GetEuler();
}
```
But then I had some problems with local/global space conversions, searched online for help and found this documentation saying that using rotations is naughty and I have to use transforms for everything :D

The root of the camera scene is a `StaticBody3D` that I don't rotate within the scene itself, I have a child node in the security camera scene that I rotate to face toward the target. However, the camera may be rotated when it is instantiated inside another scene. So, should I set global rotation of the child node to the desired rotation, or convert it to local space first?
